File: src/Zlebuh.MinTacToe.GameEngine/GameControl.cs
using Zlebuh.MinTacToe.GameModel;

namespace Zlebuh.MinTacToe.GameEngine;

public static class GameControl
{
    public static Game Initialize(Rules rules)
    {
        Grid grid = [];
        for (int row = 0; row < rules.Rows; row++)
        {
            for (int col = 0; col < rules.Columns; col++)
            {
                Coordinate coordinate = new(row, col);
                Field field = new()
                {
                    Player = null,
                    SurroundedByNotExplodedMines = 0,
                    IsMine = false,
                    Generated = false,
                    HasAllNeighboursGenerated = false
                };
                grid[coordinate] = field;
            }
        }

        Game game = new()
        {
            GameState = new()
            {
                Grid = grid,
                IsGameOver = false,
                Changes = [],
                PlayerOnTurn = Player.O,
                Winner = null,
                MovesPlayed = 0
            },
            Rules = rules
        };
        return game;
    }
}

File: packages/game-engine/src/gameControl.ts
import type { Game, Rules } from "./types.js";
import { DEFAULT_RULES } from "./types.js";
import { createGrid } from "./grid.js";

export function initialize(rules: Partial<Rules> = {}): Game {
  const resolvedRules: Rules = { ...DEFAULT_RULES, ...rules };
  return {
    rules: resolvedRules,
    gameState: {
      grid: createGrid(resolvedRules),
      isGameOver: false,
      winner: null,
      playerOnTurn: "O",
      changes: [],
      movesPlayed: 0,
    },
  };
}
